### PR TITLE
Deprecate selecting variant by configuration name for non-ivy external components

### PR DESCRIPTION
**Pull request by jvandort**
This effectively deprecates selecting variants by configuration name for maven components. Previously, users were able to select maven "configurations" such as runtime, compile, or test, using the configuration: notation when declaring the dependency.

Maven does not have the concept of configurations. Variant derivation works well for maven and we are sticking to it.

This leaves only ivy components and local components as the only components that support variant selection via configuration name. Eventually, local components will also get this same restriction. Only ivy components should allow variant-by-name selection, as that is a core aspect of its resolution model.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 1]
```diff
@@ -90,6 +90,25 @@ val myConfig = buildscript.configurations.detachedConfiguration(
...
 
+[[selecting_variant_by_configuration_name]]
+==== Selecting Maven variants by configuration name
+
+Starting in Gradle 9.0, selecting variants by name from non-ivy external components will be forbidden.
+Selecting variants by name from local components will still be permitted, however this pattern is discouraged.
+For local components, variant aware dependency resolution should be preferred over selecting variants by name.
+
+The following dependencies will fail to resolve when targeting a non-ivy external component:
+
```
Comment (big-guy) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:101]: Can we link to a section of the user manual about variant aware dependency resolution? 

```suggestion
Starting in Gradle 9.0, selecting variants by name from non-Ivy external components will be forbidden.

Selecting variants by name from local components will still be permitted; however, this pattern is discouraged.
For local components, variant aware dependency resolution should be preferred over selecting variants by name.

The following dependencies will fail to resolve when targeting a non-Ivy external component:

```

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/component/model/DefaultExternalComponentGraphResolveState.java [hunk 2]
```diff
@@ -236,10 +238,26 @@ public List<? extends VariantGraphResolveState> getVariantsForAttributeMatching(
...
+
         @Nullable
         @Override
         public VariantGraphResolveState getVariantByConfigurationName(String name) {
-            // TODO: Deprecate this method for non-ivy components.
+
+            DeprecationLogger.deprecateBehaviour("Selecting a variant by configuration name from a non-ivy external component.")
+                .willBecomeAnErrorInGradle9()
+                .withUpgradeGuideSection(8, "selecting_variant_by_configuration_name")
+                .nagUser();
```
Comment (big-guy) [inline platforms/software/dependency-management/src/main/java/org/gradle/internal/component/model/DefaultExternalComponentGraphResolveState.java:254]:
```suggestion
            DeprecationLogger.deprecateBehaviour("Selecting a variant by configuration name from a non-Ivy external component.")
                .willBecomeAnErrorInGradle9()
                .withUpgradeGuideSection(8, "selecting_variant_by_configuration_name")
                .nagUser();
```

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 3]
```diff
@@ -90,6 +90,25 @@ val myConfig = buildscript.configurations.detachedConfiguration(
...
+[source,groovy]
+----
+dependencies {
+    implementation(group: "com.example", name: "example", version: "1.0", configuration: "conf")
```
Comment (big-guy) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:105]: Can we add something to DependencyHandler docs to discourage this?

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 4]
```diff
@@ -90,6 +90,25 @@ val myConfig = buildscript.configurations.detachedConfiguration(
...
+    implementation("com.example:example:1.0") {
+        targetConfiguration = "conf"
```
Comment (big-guy) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:107]: Can we add something to ModuleDependency Javadoc to discourage this for Maven?

**Review by big-guy — CHANGES_REQUESTED** [review]

**Comment by jvandort** [comment]
@bot-gradle test this

**Pull request merged**